doc/refactoring_design.md

```diff
@@ -0,0 +1,481 @@
+# Keeper Contract Design for Tethys
+
+***DISCLAIMER: THIS IS A WORK IN PROGRESS***
+
+This document provides technical details and describes the high-level design of keeper contract in ocean protocol.
+
+## 1. Overview
+
+### 1.1 Core Modules
+
+From the top-level point of view, there are total 9 moduels as shown in the below:
+
+<img src="img/refactoring/modules.jpg" width="1000" />
+
+* **Ocean Token**: the native ERC20 token in Ocean network
+* **Exchange**: Uniswap-like exchange built inside Ocean, where user can exchange between Ether and Ocean tokens.
+* **Service Agreement**: every service is modeled as a *service agreement*. Please see [SLA.md](https://github.com/oceanprotocol/dev-ocean/blob/feature/SLA-specs/doc/sla.md) and [smart contract prototype](https://github.com/oceanprotocol/keeper-contracts/blob/develop/contracts/SLA/ServiceAgreement.sol) for details.
+* **Tribe**: it controls the registry of membership for each service agreement template; defines TCR based registries for permission, which has whitelisted members who can provide or consume the service.
+* **Curation**: users can curate a registry of high-quality service agreement templates through staking on corresponding bonding curves. It helps consumers discover high-quality assets/services. 
+* **Governance**: it governs the available assets and services in Ocean network. For example, it whitelists new service agreement templates through TCR & Voting. 
+* **Dispute Resolution**: resolves the disputes between consumer and provider through voting or TCR. 
+* **Block Reward**: Ocean network will periodically release block reward (i.e., new minted ERC20 Ocean tokens) to reward the providers of data commons. Note provider of data commons must be randomly selected to fulfill the service requests so that they have the equal probability to receive block rewards.
+* **Access Control**: it is the controller contract which grant access to consumers after depended conditions are fulfilled. The access token can be delivered in two approaches: secret store and OEP10. 
+* **Payment**: it handles all payment processing, including lock payment, release payment and refund payment. 
+
+
+### 1.2 Dependence
+
+The core modules have dependence on each other, which can be illustrated as following. Clearly, the `service agreement` module is the fundamental building block, where many other modules depends on it.  
+
+<img src="img/refactoring/dependence.jpg" width="1000" />
+
+
+## 2. Module Design
+
+In this section, each module is designed with details (some materials comes from [Research Guidance](https://docs.google.com/presentation/d/1Z6Acq2LD3eHPD1SoH_bxHH-XaOKLWemgFiKgWICekys/edit#slide=id.g44065f7377_2_1894)) and key smart contract functions are provided.
```

Comment:
Please ask @diminator if we can keep the `research guide` publicly accessible or not